Jetty 12 handler as boolean processor

gregw opened the pull request:
Another alternative Handler architecture inspired by #8978 for #8929

This PR removed the modal nature of `Request`, `Response` and `Callback`, so that they need not be accepted (via API or returning a `Processor`) before they are fully active.    Not only does this avoid a lot of locking on many methods and on the state change, but it also allows for a boolean return from `process` to no longer be a race condition.  

If a `handler.process(request,response,callback)` returns `true`, then the callback will be called and the caller must no longer act on the request/response/callback.  If `false` is returned, then the handler will never act on the request/response/callback so the caller must either handle the request itself (e.g. 404) or it is free to try another handler.
If an exception is thrown, then the caller cannot know if the callback will be called and must assume not and race with any lingering async process to write an error and complete the callback.

In #8978, I tried having a separate `Handler.Conditional` type that would have a method on it  to accept requests and only then call `process` which had to handle the request.  This is probably workable but had several cons:
 + difficult to pass state/workings/wrappers from the conditional test to the process method, so some work is done twice or we need wrappers just to carry extra state.
 + It was difficult to wrap conditional handlers with non conditional handlers.
 + callers had to do instanceof test of handlers before knowing how to call them.

So this branch is simpler, as essentially all handlers are conditional.

gregw commented on VERSION.txt:
```diff
@@ -1,77 +1,5 @@
- + 8999 Overuse of FileID.isArchive() and inability to deal with packed jars
-   without .jar extension
- + 9006 WebSocket Message InputStream read() returns signed byte
-
```
Oh no!!! my squash merge went wrong!

gregw replied:
I think I fixed this.

gregw committed:
Alternative Handler architecture

All Handlers are Processors, which now return a boolean to indicate the request has been accepted.
The request/response/callback are no longer modal, so there is no race with the boolean return.

gregw commented on documentation/jetty-asciidoctor-extensions/pom.xml:
```diff
@@ -11,10 +11,6 @@
   <name>Jetty :: Documentation :: AsciiDoctor Extensions</name>
   <packaging>jar</packaging>
 
-  <properties>
-    <bundle-symbolic-name>${project.groupId}.asciidoctor.extensions</bundle-symbolic-name>
-  </properties>
-
```
Hmm I've still missed some changes in HEAD?

gregw committed:
Alternative Handler architecture

Fixes from bad squash

gregw committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into jetty-12-handler-as-boolean-processor

janbartel commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ContextHandler.java:
```diff
@@ -653,59 +652,63 @@ public void destroy()
 
     @Override
-    public Request.Processor handle(Request request) throws Exception
+    public boolean process(Request request, Response response, Callback callback) throws Exception
```
:100: This is _so much easier_ to understand.

janbartel reviewed (approved):
I _really_ like the way this is going. The code is much simpler. It _has_ to also be faster.

gregw committed:
Alternative Handler architecture

Minor renames to reduce differences

gregw commented on jetty-ee9/jetty-ee9-nested/src/test/java/org/eclipse/jetty/ee9/nested/ContextHandlerTest.java:
```diff
@@ -757,12 +757,7 @@ public void exitScope(ContextHandler.APIContext context, Request request)
-            "EE9 exit /",
-            "Core exit http://0.0.0.0/",
-            // Enter again for process(request, response, callback)
+            // Enter for process(request, response, callback)
```
This one makes me smile :)

gregw commented on jetty-ee9/jetty-ee9-nested/src/test/java/org/eclipse/jetty/ee9/nested/ContextHandlerTest.java:
```diff
@@ -757,12 +757,7 @@ public void exitScope(ContextHandler.APIContext context, Request request)
 
         assertThat(history, contains(
-            // Enter once for handle(request)
-            "Core enter http://0.0.0.0/",
-            "EE9 enter /",
-            "EE9 exit /",
-            "Core exit http://0.0.0.0/",
-            // Enter again for process(request, response, callback)
```
This one makes me smile :)

gregw committed:
Alternative Handler architecture

Minor cleanups to reduce differences

gregw committed:
Alternative Handler architecture

Fixed race in test

gregw committed:
Alternative Handler architecture

Fixed missed merge

gregw committed:
Alternative Handler architecture

Non generic DelayedHandler

gregw committed:
Alternative Handler architecture

restored comment

gregw committed:
Merge branch 'jetty-12.0.x' into jetty-12-handler-as-boolean-processor

gregw committed:
Alternative Handler architecture

inlined the Handler.Processor class

gregw commented:
@sbordet @lorban I've been looking at redoing this branch from HEAD, and I really don't think that is the right way to go.  There is a lot of work and consideration in this branch that will take a huge effort to redo from scratch.   I know there is some risk of some changes being lost in this branch, but I honestly think that is less if we review this branch and check for them, than if we redo the changes to the many hundreds of handlers.    Whilst much of it can be done with tools, ultimately you need to visit most of the handlers to check them.  I've visited many hundreds of handlers and believe I have them mostly right.    Redoing all that work will not remove the need for a detailed review to recheck them all.

I don't think there is any significant kruft from previous branches left, and if there is it needs to be reviewed, identified and labelled as kruft anyway (rather than a change to be remade). Again, I don't think it is less work to redo the branch rather than review this one.

gregw committed:
Alternative Handler architecture

Fixed multipart test

gregw committed:
Alternative Handler architecture

Updates from review

gregw committed:
Alternative Handler architecture

Updates from review

gregw committed:
Alternative Handler architecture

fixed bad refactor

gregw committed:
Alternative Handler architecture

Fix invocation type of some handlers

gregw committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into jetty-12-handler-as-boolean-processor

# Conflicts:
#	jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/BufferedResponseHandler.java

gregw commented:
When reviewing this PR, there has been some discussion of renaming process back to handle.  I think this is a good idea, but would like to do that in a separate PR, as this is big enough as is and the faster we get it merged the better.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Handler.java:
```diff
@@ -65,78 +61,63 @@
  * {
  *     @Override
- *     public Processor handle(Request request)
+ *     public void process(Request request, Response response, Callback callback)
```
Should return `boolean`.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Handler.java:
```diff
@@ -65,78 +61,63 @@
+ * class ConditionalHandler extends Handler.Wrapper
+ * {
+ *     @Override
+ *     public void process(Request request, Response response, Callback callback)
```
Should return `boolean`.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Handler.java:
```diff
@@ -65,78 +61,63 @@
+ * }
+ * }</pre>
+ * <p>An example of a {@code Handler} that decides whether to pass the request to
+ * a child, without accepting:
```
"without accepting" should be removed.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Handler.java:
```diff
@@ -406,6 +386,22 @@ public void destroy()
             super.destroy();
         }
+
+        public abstract static class Blocking extends Abstract
```
I think this class should be removed, as `Abstract` is by default blocking.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Handler.java:
```diff
@@ -578,15 +574,14 @@ public String toString()
         }
```
Remove this `toString()` override.

sbordet commented on jetty-core/jetty-client/src/test/java/org/eclipse/jetty/client/HostnameVerificationTest.java:
```diff
@@ -72,14 +72,15 @@ public void setUp() throws Exception
                 try (Blocker.Callback blocker = Blocker.callback())
                 {
                     Content.Sink.write(response, true, "foobar", blocker);
+                    return true;
```
There is a missing call to `blocker.block()` here!

sbordet commented on jetty-core/jetty-client/src/test/java/org/eclipse/jetty/client/HttpClientProxyProtocolTest.java:
```diff
@@ -179,6 +181,8 @@ public void process(Request request, Response response, Callback callback)
                 }
                 response.getHeaders().put(HttpHeader.CONTENT_TYPE, MimeTypes.Type.TEXT_PLAIN.asString());
                 Content.Sink.write(response, true, String.valueOf(Request.getRemotePort(request)), callback);
+
```
Remove newline.

gregw replied:
REALLY!?!??!?  OK!

sbordet commented on jetty-core/jetty-client/src/test/java/org/eclipse/jetty/client/HttpClientProxyTest.java:
```diff
@@ -198,6 +198,8 @@ else if (target.startsWith("/server"))
                 {
                     Response.writeError(request, response, callback, HttpStatus.INTERNAL_SERVER_ERROR_500);
                 }
+
```
Remove newline.

gregw replied:
nah!

sbordet commented on jetty-core/jetty-client/src/test/java/org/eclipse/jetty/client/util/MultiPartRequestContentTest.java:
```diff
@@ -436,13 +438,14 @@ public void process(Request request, Response response, Callback callback) throw
             try
             {
-                process(formData.join());
+                process(formData.join()); // May block waiting for
```
```suggestion
                process(formData.join()); // May block waiting for multipart form data.
```

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Handler.java:
```diff
@@ -30,22 +29,19 @@
+ * <p>A {@code Handler} is an {@link Invocable}, so it has an
+ * {@link InvocationType}, which is by default {@link InvocationType#BLOCKING} unless a
+ * {@code NonBlocking} variant has been extended or a specific
+ * {@link InvocationType} passed to a constructor.
```
This paragraph should be moved to the `Abstract` class.
Here we should just mention that it may be blocking or non-blocking depending on the implementation.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Handler.java:
```diff
@@ -578,15 +574,14 @@ public String toString()
             }
-            return null;
+            return false;
         }
```
This class computes the InvocationType from its children, but then it does not expose it itself by overriding `getInvocationType()`, but it should.

sbordet commented on jetty-core/jetty-proxy/src/main/java/org/eclipse/jetty/proxy/ProxyHandler.java:
```diff
@@ -69,7 +69,7 @@
  * @see Reverse
  */
-public abstract class ProxyHandler extends Handler.Processor
+public abstract class ProxyHandler extends Handler.Abstract.Blocking
```
Should extend `NonBlocking`.

sbordet commented on jetty-core/jetty-rewrite/src/main/java/org/eclipse/jetty/rewrite/handler/Rule.java:
```diff
@@ -64,7 +64,7 @@ public String toString()
      * used to chain a sequence of {@link Rule}s together.
      * The tuple is initialized with only the request, then the processor is
      * then passed to a chain of rules before the ultimate processor is
```
```suggestion
     * The rule processor is initialized with the initial request, then it is
     * passed to a chain of rules before the child {@code Handler} is
```

sbordet commented on jetty-core/jetty-rewrite/src/main/java/org/eclipse/jetty/rewrite/handler/RewriteHandler.java:
```diff
@@ -103,19 +105,20 @@ public void setOriginalPathAttribute(String originalPathAttribute)
 
         // At least one rule matched, call super with the result of the rule applications.
-        return output.wrapProcessor(super.handle(output));
+        output.setProcessor(getHandler());
```
I'd prefer the child `Handler` to be passed to the constructor of `Rule.Processor`.

gregw replied:
I agree, but can we defer more changes to rewrite to another PR?

sbordet commented on jetty-core/jetty-rewrite/src/main/java/org/eclipse/jetty/rewrite/handler/Rule.java:
```diff
@@ -64,7 +64,7 @@ public String toString()
      * The tuple is initialized with only the request, then the processor is
      * then passed to a chain of rules before the ultimate processor is
-     * passed in {@link #wrapProcessor(Processor)}. Finally, the response
+     * passed in {@link #setProcessor(Processor)}. Finally, the response
```
Update this javadoc line if the child `Handler` is passed to the constructor of `RUle.Processor`.

gregw replied:
ditto

sbordet commented on jetty-core/jetty-rewrite/src/main/java/org/eclipse/jetty/rewrite/handler/Rule.java:
```diff
@@ -78,38 +78,35 @@ public Processor(Request request)
-         * @return this instance
          */
-        public Processor wrapProcessor(Processor processor)
+        public void setProcessor(Processor processor)
```
I'd prefer this to be done in the constructor, and remove this method.

gregw replied:
ditto

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Request.java:
```diff
@@ -48,26 +49,22 @@
- * public void process(Request request, Response response, Callback callback)
+ * public boolean process(Request request, Response response, Callback callback)
  * {
+ *     // reject requests not appropriate for this handler
```
```suggestion
 *     // Reject requests not appropriate for this handler.
```

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Request.java:
```diff
@@ -192,8 +189,20 @@ static String getPathInContext(Request request)
+     */
+    default long getNanoTimeStamp()
+    {
+        return System.nanoTime() - TimeUnit.MILLISECONDS.toNanos(System.currentTimeMillis() - getTimeStamp());
```
We have introduced a `NanoTime` class, so please never use `System.nanoTime()` ever again.

This method should be renamed to just `getNanoTime()`, as we have standardized to that naming.

Plus the implementation should really compare against an initial nanoTime taken at creation time, because `System.currentTimeMillis()` can be adjusted by NTP and this method may return negative times, which would be counter-intuitive for the typical nanoTime usage.

Finally, this method is never used, not even by the `StatisticsHandler`.
Why was it added?

gregw replied:
It probably should be used by the stats handler, although that is probably getting it directly from the nano timestamp in the HttpStream.
When we really have a channel/stream, then the actual implementation is :
```java

            HttpStream stream = _httpChannel.getHttpStream();
            if (stream != null)
                return stream.getNanoTimeStamp();
```

This in turn calls `NanoTime.now()`. So this impl is only for test requests.

The method exists because we want to encourage use of nano time for latencies measured from the first time the request is received.   It is different to the `getTimeStamp` method which is returning the wall clock.

The stats handler should use this rather than create its own timestamp that is someway through the handler chain.

I'll improve javadoc.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Request.java:
```diff
@@ -233,7 +242,7 @@ default void push(MetaData.Request resource)
     TunnelSupport getTunnelSupport();
 
-    void addHttpStreamWrapper(Function<HttpStream, HttpStream.Wrapper> wrapper);
+    void addHttpStreamWrapper(Function<HttpStream, HttpStream> wrapper);
```
Not sure about this change? What's the reason behind it?
I don't dislike because it's a bit more generic, but still?

gregw replied:
I don't recall why I  did it.  I can take it or leave it.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Request.java:
```diff
@@ -459,38 +468,43 @@ static String toRedirectURI(Request request, String location)
-         * <p>The processing may be asynchronous, i.e. this method may return early and
-         * complete the given callback later, possibly from a different thread.</p>
+         * <p>Invoked to process the passed HTTP request and response.</p>
+         * <p>The the request is accepted by returning true, then processing <em>must</em> be concluded by
```
Double "The" at beginning of sentence, plus the sentence does not read well. 
Did you mean "If the request is accepted..."?

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Request.java:
```diff
@@ -459,38 +468,43 @@ static String toRedirectURI(Request request, String location)
+         * then the callback must not be invoked and any mutation on the response reversed.</p>
          * <p>Within an implementation of this method it is possible to read the
-         * request content (that was forbidden in {@link Handler#handle(Request)}.</p>
+         * request content (that was forbidden in {@link Handler#process(Request, Response, Callback)}.</p>
```
Reading is not forbidden anymore.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Request.java:
```diff
@@ -659,55 +679,6 @@ static long getContentBytesRead(Request request)
-                processor.process(this, response, callback);
-        }
-    }
-
```
😢

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/BufferedResponseHandler.java:
```diff
@@ -139,15 +137,16 @@ public Request.Processor handle(Request request) throws Exception
         }
 
         // If not a supported path this URI is always excluded.
+        final String path = Request.getPathInContext(request);
```
Remove `final`.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/BufferedResponseHandler.java:
```diff
@@ -160,16 +159,14 @@ public Request.Processor handle(Request request) throws Exception
-            BufferedResponse bufferedResponse = new BufferedResponse(rq, rs, callback);
-            processor.process(rq, bufferedResponse, bufferedResponse);
-        };
+        // Install buffered interceptor and handle.
```
```suggestion
        // Install buffered wrappers and forward processing.
```

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ContextHandler.java:
```diff
@@ -653,59 +652,63 @@ public void destroy()
+        String path = request.getHttpURI().getCanonicalPath();
+        String pathInContext = _context.getPathInContext(path);
+        if (pathInContext == null)
+            return false;
```
The logic to match the path is completely different now.
Care to explain why, and whether this new one replaces the old logic (to make sure it's not a merge mistake or similar)?

gregw replied:
All the "path startswith context paths"  checks are now always done in _context.getPathInContext(path), which was being called in the previous impl anyway (inside processByContextHandler).
Now that handle and process are not separated, no need to do those checks twice.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ContextHandler.java:
```diff
@@ -653,59 +652,63 @@ public void destroy()
-        return contextRequest.wrapProcessor(processor);
+    protected boolean processByContextHandler(String pathInContext, ContextRequest request, Response response, Callback callback)
+    {
+        return false;
```
The whole implementation of this method was deleted?

gregw replied:
Yep, it is all done in processMovedPermanently now

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ContextHandlerCollection.java:
```diff
@@ -323,20 +313,6 @@ private Mapping(List<Handler> handlers, Map<String, Branch[]> path2Branches)
-                    }
-                }
-            }
-            _contextBranches = Collections.unmodifiableMap(contextBranches);
```
Ok for this code to be deleted?

gregw replied:
It was already the case that _contextBranches is never used.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ContextRequest.java:
```diff
@@ -16,52 +16,21 @@
-        finally
-        {
-            // We exit scope here, even though process is asynchronous, as we have wrapped
-            // all our callbacks to re-enter the scope.
```
Please retain this comment in `ContextHandler` where this code was moved to?

gregw replied:
it is in ContextHandler. comment moved.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/DebugHandler.java:
```diff
@@ -34,77 +37,41 @@
-    {
-        final Response base_response = baseRequest.getResponse();
         final Thread thread = Thread.currentThread();
         final String old_name = thread.getName();
```
Remove `final`.

lorban commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Handler.java:
```diff
@@ -406,6 +386,22 @@ public void destroy()
             super.destroy();
         }
+
+        public abstract static class Blocking extends Abstract
```
Since you introduced both `Blocking` and `NonBlocking` then `Abstract` shouldn't have any default. Possibly it shouldn't even be visible.

gregw replied:
I've removed Blocking

lorban commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Handler.java:
```diff
@@ -30,22 +29,19 @@
+ * and/or to intercept the read of the request content; and/or intercept the generation of the
+ * response; and/or to intercept the completion of the callback.
+ * <p>A {@code Handler} is an {@link Invocable}, so it has an
+ * {@link InvocationType}, which is by default {@link InvocationType#BLOCKING} unless a
```
The `InvocationType` side isn't super clear.
I'm not sure what being blocking by default means here, since it depends on the `Handler` subclass that is inherited.

gregw replied:
It means that unless the handler says otherwise, the caller of `process` will assume that it blocks

lorban commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Request.java:
```diff
@@ -192,8 +189,20 @@ static String getPathInContext(Request request)
+     */
+    default long getNanoTimeStamp()
+    {
+        return System.nanoTime() - TimeUnit.MILLISECONDS.toNanos(System.currentTimeMillis() - getTimeStamp());
```
The ms accuracy makes this impl useless, and emulating elapsed time with wall-clock time is just broken.

gregw replied:
This is just a default impl, mostly used for tests.  The real impl uses the proper nano timestamp from the HttpStream

lorban commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/DefaultHandler.java:
```diff
@@ -52,7 +52,7 @@
  * For all other requests a normal 404 is served.
  */
-public class DefaultHandler extends Handler.Processor
+public class DefaultHandler extends Handler.Abstract.Blocking
```
Is it blocking?

lorban commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/MovedContextHandler.java:
```diff
@@ -147,10 +147,10 @@ public void setCacheControl(String cacheControl)
     }
 
-    private class Redirector extends Handler.Processor
+    private class Redirector extends Blocking
```
Is it blocking?

sbordet replied:
Either `Abstract` or better yet `Abstract.NonBlocking`.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/DelayedHandler.java:
```diff
@@ -34,185 +32,202 @@
+
+        delay(delayed);
+        return true;
+    }
```
I'd like to propose an alternative implementation.

`newDelayProcess()` becomes `boolean canDelay(request, response, callback)`.
In this way there is no need to allocate if the handler cannot delay (e.g. not a multipart, etc.)

`delay()` becomes `void delay(request, response, callback)`.
In this way, each subclass can create its own private record to wrap the 3 parameters, and arrange the delayed processing.
There will be no explicit need of base class `DelayedProcess`, and even if kept for convenience it may be `private` rather than `public`.

gregw replied:
Can we move this to a different PR. I already have a bunch of changes that I'd like to make.   I'll open a draft PR against this branch to discuss

gregw replied:
Let's experiment in #9056

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/DelayedHandler.java:
```diff
@@ -34,185 +32,202 @@
+                return null;
+
+            // TODO: add logic to not delay if it's a CONNECT request.
+            // TODO: also add logic to not delay if it's a request that expects 100 Continue.
```
If my alternative implementation is accepted, then this comment applies to all delayings, so it should be moved to just before the call to `canDelay(...)`, or maybe inside `canDelay()` (subclasses need to call `super` then).

gregw replied:
Let's experiment in #9056

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ErrorProcessor.java:
```diff
@@ -115,9 +115,10 @@ public void process(Request request, Response response, Callback callback)
             {
                 // TODO: cannot write the error response, give up and close the stream.
-                x.printStackTrace();
+                LOG.warn("Cannot generate error", x);
```
```suggestion
                LOG.warn("Failure while generating error response", x);
```

lorban commented on jetty-ee9/jetty-ee9-servlets/src/test/java/org/eclipse/jetty/ee9/servlets/ThreadStarvationTest.java:
```diff
@@ -225,17 +225,18 @@ public boolean flush(ByteBuffer... buffers) throws IOException
 
             AtomicInteger count = new AtomicInteger(0);
-            _server.setHandler(new Handler.Processor()
+            _server.setHandler(new Handler.Abstract()
```
It should inherit from `Handler.Abstract.Blocking` IMHO.

lorban commented on jetty-ee9/jetty-ee9-servlet/src/test/java/org/eclipse/jetty/ee9/servlet/GzipHandlerIsHandledTest.java:
```diff
@@ -115,10 +117,10 @@ public EventHandler(LinkedBlockingQueue<String> events, String action)
         {
             events.offer(action);
-            return null;
+            return true;
```
Shouldn't this return `false`?

gregw replied:
I think it should succeed the callback.

lorban commented on jetty-ee9/jetty-ee9-proxy/src/test/java/org/eclipse/jetty/ee9/proxy/ForwardProxyTLSServerTest.java:
```diff
@@ -880,10 +874,10 @@ private static void sleep(long ms)
     }
 
-    private static class ServerHandler extends Handler.Processor
+    private static class ServerHandler extends Handler.Abstract.Blocking
```
Is it blocking?

lorban commented on jetty-ee9/jetty-ee9-proxy/src/test/java/org/eclipse/jetty/ee9/proxy/ForwardProxyServerTest.java:
```diff
@@ -233,14 +233,15 @@ public void testIPv6WithXForwardedForHeader(String ipv6) throws Exception
         httpConfig.addCustomizer(new ForwardedRequestCustomizer());
         ConnectionFactory http = new HttpConnectionFactory(httpConfig);
-        startServer(null, http, new Handler.Processor()
+        startServer(null, http, new Handler.Abstract()
```
This should inherit from `Handler.Abstract.NonBlocking`, should it?

gregw replied:
yeah, but I don't think we need to be too rigorous with our test handlers

lorban commented on jetty-ee9/jetty-ee9-proxy/src/test/java/org/eclipse/jetty/ee9/proxy/EmptyServerHandler.java:
```diff
@@ -18,10 +18,11 @@
 import org.eclipse.jetty.util.Callback;
 
-public class EmptyServerHandler extends Handler.Processor
+public class EmptyServerHandler extends Handler.Abstract.Blocking
```
Is this blocking?

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/InetAccessHandler.java:
```diff
@@ -203,13 +207,16 @@ public void excludeConnectors(String... names)
-        return super.handle(request);
+        {
+            Response.writeError(request, response, callback, 403);
+            return true;
```
Returning `null` before would have resulted in a 404, so maybe 403 is better here, but can you use `HttpStatus` constant as they have in their name a description of what 403 is (it's "forbidden", but I had to look it up).

gregw replied:
OK.  I think a false return might be better.  I've put a TODO in for now.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ResourceHandler.java:
```diff
@@ -132,19 +134,22 @@ protected ResourceService.WelcomeFactory setupWelcomeFactory()
 
-        return (rq, rs, cb) -> _resourceService.doGet(rq, rs, cb, content);
+        _resourceService.doGet(request, response, callback, content);
+        return true;
```
I don't think we want to keep this double call style to `_resourceService`.

The whole point of having `getContent()` and `doGet()` separated was that it was necessary to check whether there was content to avoid returning a processor.

Now we don't need to do that anymore, so we can go back to Jetty 11's style by just doing:

`return _resourceService.process(request, response, callback)`

with the same semantic of "if you return true you must complete the callback" and if there is no content, return false.

But perhaps in another PR?

gregw replied:
Sure, I think there is still a lot of resource service cleanup to be done after this PR.  @lachlan-roberts keep a tentacle on this :)

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/StatisticsHandler.java:
```diff
@@ -249,18 +244,31 @@ public double getProcessingTimeStdDev()
         private StatisticsRequest(Request request)
         {
             super(request);
+            _startNanoTime = System.nanoTime();
```
Use `NanoTime.now()` instead.

gregw replied:
I'm using that now, but we also need to review if a new timestamp is needed or if request.getRequestNanoTimestamp() should be used instead.  potentially both are required.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/StatisticsHandler.java:
```diff
@@ -249,18 +244,31 @@ public double getProcessingTimeStdDev()
-        private final LongAdder _bytesWritten = new LongAdder();
-        private long _processStartNanoTime;
+        protected final LongAdder _bytesRead = new LongAdder();
+        protected final LongAdder _bytesWritten = new LongAdder();
```
Restore `private` and add a `long` getter.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/StatisticsHandler.java:
```diff
@@ -249,18 +244,31 @@ public double getProcessingTimeStdDev()
     }
 
-    private class StatisticsRequest extends Request.WrapperProcessor
+    protected class StatisticsRequest extends Request.Wrapper implements Connection.Listener
```
Restore `private`.

gregw replied:
hmmm ok... but can we do another review of the statsHandler.  It needs work regardless.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/StatisticsHandler.java:
```diff
@@ -277,98 +285,67 @@ public Object getAttribute(String name)
         }
 
-        private long spentTimeNs()
+        protected long spentTimeNs()
```
Restore `private`.

gregw replied:
ditto

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/StatisticsHandler.java:
```diff
@@ -31,61 +32,91 @@
+
+            if (next.process(statisticsRequest, response, callback))
+                return true;
+            _requestStats.decrement();
```
I would add a comment saying that if the processing is done, then `_requestStats` is decremented elsewhere, because doing it after `return true` looks like a bug.

gregw replied:
ditto

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/StatisticsHandler.java:
```diff
@@ -249,18 +244,31 @@ public double getProcessingTimeStdDev()
+            _connectionStats.remove(getConnectionMetaData().getId());
+        }
+
+        public HttpStream asHttpStream(HttpStream httpStream)
```
Rename to `wrapHttpStream`.

gregw replied:
ditto

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ThreadLimitHandler.java:
```diff
@@ -156,54 +160,30 @@ public void exclude(String inetAddressPattern)
-            // if disabled, handle normally
-            return baseProcessor;
+            next.process(request, response, callback);
+            return false;
```
Should not this be this instead?

```java
return next.process(request, response, callback);
```

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ThreadLimitHandler.java:
```diff
@@ -156,54 +160,30 @@ public void exclude(String inetAddressPattern)
+        // We accept the request and will always process it.
+        LimitedRequest limitedRequest = new LimitedRequest(remote, next, request, response, callback);
+        limitedRequest.process();
+        return true;
```
Feels a lot like a `DelayedHandler`, I wonder why it is not?

gregw replied:
Because it is not just delaying dispatch, but demand on write callbacks as well.
But yeah closely related.  Maybe something to consider in #9056 ?

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ThreadLimitHandler.java:
```diff
@@ -303,6 +283,196 @@ private String getXForwardedFor(Request request)
+            try
+            {
+                if (!_handler.process(this, _response, callback))
+                    Response.writeError(this, _response, callback, 404);
```
Use `HttpStatus` instead of `404`.

gregw replied:
ack.  10 other examples

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ThreadLimitHandler.java:
```diff
@@ -156,54 +160,30 @@ public void exclude(String inetAddressPattern)
+        return true;
     }
 
     private static void getAndClose(CompletableFuture<Closeable> cf)
```
This method is not supposed to block calling `cf.get()`, but it always gets a completed `cf`, so perhaps it's best to call `getNow(null)`.

gregw replied:
... and use IO.close

lorban commented on jetty-ee9/jetty-ee9-proxy/src/test/java/org/eclipse/jetty/ee9/proxy/ClientAuthProxyTest.java:
```diff
@@ -134,17 +134,18 @@ private void startServer(Handler handler) throws Exception
     private void startServer() throws Exception
     {
-        startServer(new Handler.Processor()
+        startServer(new Handler.Abstract()
```
Should it not inherit from `Handler.Abstract.NonBlocking`?

lorban commented on jetty-ee9/jetty-ee9-nested/src/test/java/org/eclipse/jetty/ee9/nested/HttpServerTestFixture.java:
```diff
@@ -87,31 +87,33 @@ protected void startServer(org.eclipse.jetty.ee9.nested.Handler handler) throws
     }
 
-    protected static class OptionsHandler extends org.eclipse.jetty.server.Handler.Processor
+    protected static class OptionsHandler extends Handler.Abstract.Blocking
```
Should it not inherit from `Handler.Abstract.NonBlocking`?

lorban commented on jetty-ee9/jetty-ee9-nested/src/test/java/org/eclipse/jetty/ee9/nested/HttpServerTestFixture.java:
```diff
@@ -87,31 +87,33 @@ protected void startServer(org.eclipse.jetty.ee9.nested.Handler handler) throws
     }
 
-    protected static class HelloWorldHandler extends org.eclipse.jetty.server.Handler.Processor
+    protected static class HelloWorldHandler extends Handler.Abstract.Blocking
```
Should it not inherit from `Handler.Abstract.NonBlocking`?

lorban commented on jetty-ee9/jetty-ee9-nested/src/test/java/org/eclipse/jetty/ee9/nested/HttpServerTestFixture.java:
```diff
@@ -87,31 +87,33 @@ protected void startServer(org.eclipse.jetty.ee9.nested.Handler handler) throws
     }
 
-    protected static class SendErrorHandler extends org.eclipse.jetty.server.Handler.Processor
+    protected static class SendErrorHandler extends Handler.Abstract.Blocking
```
Should it not inherit from `Handler.Abstract.NonBlocking`?

lorban commented on jetty-ee9/jetty-ee9-nested/src/test/java/org/eclipse/jetty/ee9/nested/HttpServerTestFixture.java:
```diff
@@ -179,26 +182,28 @@ public void process(org.eclipse.jetty.server.Request request, org.eclipse.jetty.
     }
 
-    protected static class ReadHandler extends org.eclipse.jetty.server.Handler.Processor
+    protected static class ReadHandler extends Handler.Abstract.Blocking
```
Should it not inherit from `Handler.Abstract.NonBlocking`?

lorban commented on jetty-ee9/jetty-ee9-nested/src/test/java/org/eclipse/jetty/ee9/nested/AbstractHttpTest.java:
```diff
@@ -102,7 +102,7 @@ public TestCommitException()
     }
 
-    protected class ThrowExceptionOnDemandHandler extends Handler.Processor
+    protected class ThrowExceptionOnDemandHandler extends Handler.Abstract.Blocking
```
Should it not inherit from `Handler.Abstract.NonBlocking`?

lorban reviewed (changes requested):
Overall, I like the change from the processor lambda to a single process method returning a boolean, once you get over the fact that you can manipulate the request/response/callback before you return true the API is much simpler than the handler/processor one.

But there's some confusion left around the many `Handler` implementations: `Handler.Abstract` vs `Handler.Wrapper` vs `Handler.Abstract.Blocking` vs `Handler.Abstract.NonBlocking` is not entirely clear in my mind and that makes me question the choice of blocking vs non-blocking invocation and why `Handler.Abstract` is no different than `Handler.Abstract.Blocking`. This feels like it needs some cleanup such as a more readable API is exposed or some better doc. This could be done in an upcoming PR though, as the current one already is loaded.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ThreadLimitHandler.java:
```diff
@@ -303,6 +283,196 @@ private String getXForwardedFor(Request request)
+                @Override
+                public void succeeded()
+                {
+                    CompletableFuture<Closeable> futurePermit = _remote.acquire();
```
I don't think this is correct in general.
If we are writing synchronously and the callback is succeeded synchronously, we have already acquired a permit to process the request, but here we may fail to acquire the same permit (although we do have a permit already).

gregw replied:
Oh yeah... we need a "do you already have a permit" semantic.  I'll add a TODO and we can work on it in #9056

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ThreadLimitHandler.java:
```diff
@@ -303,6 +283,196 @@ private String getXForwardedFor(Request request)
+                @Override
+                public void failed(Throwable x)
+                {
+                    CompletableFuture<Closeable> futurePermit = _remote.acquire();
```
Ditto.

gregw replied:
Ditto

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/TryPathsHandler.java:
```diff
@@ -126,17 +132,21 @@ public void setPaths(List<String> paths)
-            if (childProcessor != null)
-                return result.wrapProcessor(childProcessor);
+            if (!super.process(new TryPathsRequest(request, interpolate(request, paths.get(0))), response, callback))
+                Response.writeError(request, response, callback, 404);
```
Use `HttpStatus`.

gregw replied:
ditto

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/TryPathsHandler.java:
```diff
@@ -145,6 +155,65 @@ private String interpolate(Request request, String value)
         return value.replace("$path", path);
     }
 
+    private class PathsIterator extends IteratingCallback
```
Extend from `IteratingNestedCallback` instead, so you can avoid to override `onCompleteSuccess()` and `onCompleteFailure()` below.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/TryPathsHandler.java:
```diff
@@ -145,6 +155,65 @@ private String interpolate(Request request, String value)
+        private final Response response;
+        private final Callback callback;
+        private final Iterator<String> paths = TryPathsHandler.this.paths.iterator();
+        boolean trying;
```
Make it `private`.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/TryPathsHandler.java:
```diff
@@ -126,17 +132,21 @@ public void setPaths(List<String> paths)
     @Override
-    public Request.Processor handle(Request request) throws Exception
+    public boolean process(Request request, Response response, Callback callback) throws Exception
     {
```
We are missing the check about the next handler:

```java
Handler next = getHandler();
if (next == null)
    return false;
```

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/TryPathsHandler.java:
```diff
@@ -145,6 +155,65 @@ private String interpolate(Request request, String value)
+        {
+            if (trying)
+            {
+                if (response.getStatus() != 404)
```
Use `HttpStatus`.
I would be more strict and only accept `HttpStatus.OK_200`.

gregw replied:
Reaaly?  There are lots of valid 2XX, 3XX responses that could come back, e.g. for static content not modified

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/TryPathsHandler.java:
```diff
@@ -145,6 +155,65 @@ private String interpolate(Request request, String value)
+                    return Action.SUCCEEDED;
+                trying = false;
+                response.reset();
+                response.setStatus(200);
```
Use `HttpStatus`.
But why are we setting it to `200`, I think it should stay at `0` as if it was the first path, and that is already done by `response.reset()`.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/TryPathsHandler.java:
```diff
@@ -184,4 +253,21 @@ public HttpURI getHttpURI()
+        public void write(boolean last, ByteBuffer byteBuffer, Callback callback)
+        {
+            if (getStatus() == HttpStatus.NOT_FOUND_404)
+                getRequest().getComponents().getThreadPool().execute(callback::succeeded);
```
Why executing `callback.succeeded()`?

gregw replied:
To avoid deep recursion

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/gzip/GzipHandler.java:
```diff
@@ -521,127 +513,86 @@ public void setInflateBufferSize(int size)
-        // Update headers for etags and inflation
-        HttpFields.Mutable newFields = null;
-        if (inflated || httpFields.contains(ETAG_HEADERS))
+        final String path = Request.getPathInContext(request);
```
Remove `final`.

gregw replied:
I'm not sure what you have against final variables..... but OK.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/gzip/GzipHandler.java:
```diff
@@ -521,127 +513,86 @@ public void setInflateBufferSize(int size)
-        HttpFields.Mutable newFields = null;
-        if (inflated || httpFields.contains(ETAG_HEADERS))
+        final String path = Request.getPathInContext(request);
+        boolean tryInflate = getInflateBufferSize() >= 0 && isPathInflatable(path);
```
Rename to `isPathDeflatable()`.

gregw replied:
yeah it is a better name, but it just creates more naming inconsistencies whilst this is a GzipHandler and can only do gzip.  We need other PRs on this class.   But I'll do these renames

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/gzip/GzipHandler.java:
```diff
@@ -521,127 +513,86 @@ public void setInflateBufferSize(int size)
+        return false;
+    }
+
+    protected boolean isPathMimeTypeGzipable(MimeTypes mimeTypes, String requestURI)
```
Rename to `isMimeTypeDeflatable()`.

sbordet replied:
This was not renamed?

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/gzip/GzipResponseAndCallback.java:
```diff
@@ -77,14 +79,51 @@ public class GzipResponse extends Response.Wrapper
     private ByteBuffer _buffer;
 
-    public GzipResponse(Request request, Response wrapped, GzipFactory factory, HttpField vary, int bufferSize, boolean syncFlush)
+    public GzipResponseAndCallback(Request request, Response wrapped, Callback callback, GzipHandler handler)
```
Elsewhere, `request, response, callback` were always the last 3 parameters, so please move `GzipHandler` as first parameter.

gregw replied:
We are a bit inconsistent on this.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/gzip/GzipResponseAndCallback.java:
```diff
@@ -122,8 +161,9 @@ protected void commit(boolean last, Callback callback, ByteBuffer content)
         // Are we excluding because of status?
-        Response response = GzipResponse.this;
+        Response response = GzipResponseAndCallback.this;
         Request request = response.getRequest();
```
```suggestion
        Request request = getRequest();
        Response response = this;
```

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/gzip/GzipResponseAndCallback.java:
```diff
@@ -390,7 +430,7 @@ private Action compressing(Deflater deflater, ByteBuffer outputBuffer)
         /**
          * This method is called by {@link #compressing(Deflater, ByteBuffer)}, once the last chunk is compressed;
-         * or directly from {@link #process()} if an earlier call to {@link #finishing(Deflater, ByteBuffer)} finishing was unable to complete.
+         * or directly from {@link #process()} if an earlier call to {@code finishing} was unable to complete.
```
```suggestion
         * or directly from {@link #process()} if an earlier call to this method was unable to complete.
```

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:
```diff
@@ -590,15 +550,15 @@ public void run()
             Server server = _connectionMetaData.getConnector().getServer();
 
-            Request.Processor processor;
             Request customized = request;
```
Move the definition of this variable closer to where the customizers are called.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:
```diff
@@ -633,52 +592,58 @@ public void run()
-                if (processor == null)
-                    processor = (req, res, cb) -> Response.writeError(req, res, cb, HttpStatus.NOT_FOUND_404);
+                if (!server.process(customized, request._response, request._callback))
+                    Response.writeError(customized, request._response, request._callback, 404);
```
Use `HttpStatus`.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:
```diff
@@ -633,52 +592,58 @@ public void run()
-            {
-                _processState = ProcessState.PROCESSING;
-            }
+            boolean accepted;
```
Remove this variable, unused.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:
```diff
@@ -633,52 +592,58 @@ public void run()
-                _processState = ProcessState.PROCESSED;
-
-                if (_failure != null)
+                if (failure == null)
```
This is always `false` because if `complete==false` then `failure!=null` from the logic above.
I'm worried the original logic was not ported over properly?

gregw replied:
I've cleaned it up a bit, but let's walk through it together.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:
```diff
@@ -633,52 +592,58 @@ public void run()
 
-                completeStream = _completed && (failure != null || _writeState == WriteState.LAST_WRITE_COMPLETED);
-                stream = _stream;
+                if (failure != null)
```
This expression is always `true`, see previous comment.

gregw replied:
ditto

sbordet commented on jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/HalfCloseTest.java:
```diff
@@ -159,7 +159,7 @@ public int getHandled()
-    public static class AsyncHandler extends Handler.Processor
+    public static class AsyncHandler extends Handler.Abstract.Blocking
     {
         transient int handled;
```
Remove `transient` and make it `private`.

sbordet commented on jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/HttpServerTestBase.java:
```diff
@@ -1563,24 +1564,26 @@ public void testWriteBodyAfterNoBodyResponse() throws Exception
         while (line != null);
```
Convert this to a normal while loop:

```java
while (line != null)
{
    line = in.readLine();
}
```

sbordet commented on jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/handler/DelayedHandlerTest.java:
```diff
@@ -213,87 +213,24 @@ public void process(Request request, Response response, Callback callback) throw
-            String content = new String(response.getContentBytes(), StandardCharsets.UTF_8);
-            assertThat(content, containsString("Hello"));
-        }
-    }
```
Why was removed?

gregw replied:
Because I removed the QoS varient of the Delayed handler.  IT was kind of the same thing as ThreadLimitHandler, but also not implemented completely.    We can review in #9056

sbordet commented on jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/handler/DelayedHandlerTest.java:
```diff
@@ -332,78 +269,22 @@ public Request.Processor handle(Request request)
-            String content = new String(response.getContentBytes(), StandardCharsets.UTF_8);
-            assertThat(content, containsString("<p>No context on this server matched or handled this request.</p>"));
-        }
-    }
```
Why was removed?

gregw replied:
tested already in testDelayed404.  Can review in #9056

sbordet commented on jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/handler/StatisticsHandlerTest.java:
```diff
@@ -459,214 +462,82 @@ public void testTwoRequestsInParallel() throws Exception
-//        assertThat(_statsHandler.getDispatchedTimeTotal(), greaterThanOrEqualTo(dispatchTime * 2 * 3 / 4));
-//        assertTrue(_statsHandler.getDispatchedTimeMean() + dispatchTime <= _statsHandler.getDispatchedTimeTotal());
-//        assertTrue(_statsHandler.getDispatchedTimeMax() + dispatchTime <= _statsHandler.getDispatchedTimeTotal());
-//    }
```
Why was removed? Should not be restored or at least kept?

gregw replied:
Some tests were no longer meaningful once there were no longer two phases.
I think stats handler needs more work and separate review.

sbordet commented on jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/handler/StatisticsHandlerTest.java:
```diff
@@ -675,372 +546,151 @@ public Request.Processor handle(Request request)
-//        assertEquals(_statsHandler.getRequestTimeTotal(), _statsHandler.getRequestTimeMean(), 0.01);
-//
-//        assertThat(_statsHandler.getDispatchedTimeTotal(), greaterThanOrEqualTo(dispatchTime * 3 / 4));
-//    }
```
Why removed?

gregw replied:
Some tests were no longer meaningful once there were no longer two phases.
I think stats handler needs more work and separate review.

sbordet commented on jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/handler/TryPathsHandlerTest.java:
```diff
@@ -169,7 +160,7 @@ public void process(Request request, Response response, Callback callback)
 
     @Test
-    public void testTryPathsWithPathMappings() throws Exception
+    public void testTryPathsAsync404() throws Exception
```
Please restore the original test, which mimics what we have on the website and what one should do to have a proper behavior for PHP forwarding.

sbordet commented on jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/handler/gzip/GzipHandlerTest.java:
```diff
@@ -1995,20 +2003,13 @@ public void process(Request request, Response response, Callback callback) throw
         @Override
-        public Request.Processor handle(Request request) throws Exception
+        public boolean process(Request request, Response response, Callback callback) throws Exception
         {
```
Missing the check on the next `Handler`:

```java
Handler next = getHandler();
if (handler == null)
    return false;
```

gregw replied:
It is a test handler and it always has a nested handler

sbordet commented on jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/ssl/SniSslConnectionFactoryTest.java:
```diff
@@ -100,22 +100,16 @@ private void start(BiConsumer<SslContextFactory.Server, SecureRequestCustomizer>
             @Override
-            public Request.Processor handle(Request request) throws Exception
+            public boolean process(Request request, Response response, Callback callback) throws Exception
             {
```
Missing the check on the next `Handler`:

```java
Handler next = getHandler();
if (handler == null)
    return false;
```

gregw replied:
ditto

sbordet commented on jetty-core/jetty-session/src/test/java/org/eclipse/jetty/session/SimpleSessionHandler.java:
```diff
@@ -60,16 +60,15 @@ public Server getServer()
+        SessionRequest sessionRequest = new SessionRequest(request);
         addSessionStreamWrapper(request);
-        return sessionRequest.wrapProcessor(processor);
+        return sessionRequest.process(next, response, callback);
```
No need to pass the `Handler`, since `SessionRequest` is not `static`.

Also, I would use the same style used elsewhere, where `request, response, callback` are passed to the constructor (so they can be `final`), and a parameter-less `process()` method.

gregw replied:
That style is useful if the response and callback need to be remembered, hence save them as final fields.  But they are used only within the scope of the process method.

sbordet commented on jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/security/IdentityService.java:
```diff
@@ -28,7 +29,7 @@
      * Associate a user identity with the current thread.
      * This is called with as a thread enters the
-     * {@link SecurityHandler#handle(Request)}
+     * {@link Handler#process(Request, org.eclipse.jetty.server.Response, org.eclipse.jetty.util.Callback)}
```
Previous javadoc referred to `SecurityHandler`, so perhaps just refer to the class, rather than the method?

gregw replied:
I find the method clearer as to what is being entered.  Note that we may need to associate on demand and write callbacks? I've added TODO

sbordet commented on jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/security/SecurityHandler.java:
```diff
@@ -448,144 +448,135 @@ public void setSessionRenewedOnAuthentication(boolean renew)
         
         final ServletContextRequest servletContextRequest = Request.as(request, ServletContextRequest.class);
-        final ServletContextRequest.ServletApiRequest servletApiRequest =
-            (servletContextRequest == null ? null : servletContextRequest.getServletApiRequest());
+        if (servletContextRequest == null)
+            return false;
+        final ServletContextRequest.ServletApiRequest servletApiRequest = servletContextRequest.getServletApiRequest();
         final Authenticator authenticator = _authenticator;
```
Remove `final`.

gregw replied:
I like them

sbordet commented on jetty-ee9/jetty-ee9-demos/jetty-ee9-demo-embedded/src/main/java/org/eclipse/jetty/ee9/demos/HelloHandler.java:
```diff
@@ -47,20 +43,16 @@ public HelloHandler(String greeting, String body)
-        };
-
+        else
+            response.write(true, BufferUtil.toBuffer("<h1>" + greeting + "</h1>"),
```
This first write should have `last=false`.

sbordet commented on jetty-ee9/jetty-ee9-demos/jetty-ee9-demo-embedded/src/main/java/org/eclipse/jetty/ee9/demos/ManyHandlers.java:
```diff
@@ -81,15 +80,11 @@ public Request.Processor handle(Request request) throws Exception
         @Override
-        public Request.Processor handle(Request request) throws Exception
+        public boolean process(Request request, Response response, Callback callback) throws Exception
         {
```
Missing null check:

```java
Handler next = getHandler();
if (handler == null)
    return false;
```

sbordet commented on jetty-ee9/jetty-ee9-nested/src/test/java/org/eclipse/jetty/ee9/nested/AsyncCompletionTest.java:
```diff
@@ -116,11 +117,8 @@ protected void startServer(Handler handler) throws Exception
             @Override
-            public org.eclipse.jetty.server.Request.Processor handle(org.eclipse.jetty.server.Request request) throws Exception
+            public boolean process(org.eclipse.jetty.server.Request request, Response response, Callback callback) throws Exception
             {
```
Missing null check:

```java
Handler next = getHandler();
if (handler == null)
    return false;
```

gregw replied:
just a test

sbordet commented on jetty-ee9/jetty-ee9-proxy/src/test/java/org/eclipse/jetty/ee9/proxy/EmptyServerHandler.java:
```diff
@@ -18,10 +18,11 @@
-    public void process(Request request, Response response, Callback callback)
+    public boolean process(Request request, Response response, Callback callback)
     {
+        return true;
```
The `callback` must be succeeded.

gregw committed:
Optimize PathMappings

Avoid iterations if only ServletPathSpec instances
Avoid tests for empty mappings.
Better reset implementation

gregw committed:
Optimize PathMappings

Avoid iterations if only ServletPathSpec instances
Avoid tests for empty mappings.
Better reset implementation

gregw committed:
Removed Abstract.Blocking

gregw committed:
updates from @lorban review

gregw committed:
Update jetty-core/jetty-rewrite/src/main/java/org/eclipse/jetty/rewrite/handler/Rule.java

Co-authored-by: Simone Bordet <simone.bordet@gmail.com>

gregw committed:
updates from @sbordet review

gregw committed:
Merge branch 'jetty-12-handler-as-boolean-processor' of https://github.com/eclipse/jetty.project into jetty-12-handler-as-boolean-processor

gregw committed:
updates from @sbordet review

gregw committed:
updates from @sbordet review

gregw committed:
updates from @sbordet review

gregw committed:
updates from @sbordet review

gregw committed:
updates from @sbordet review

gregw committed:
Optimize PathMappings

Improve suffix matching

gregw committed:
Optimize PathMappings

Improve exact matching

gregw committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into jetty-12-handler-as-boolean-processor

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Request.java:
```diff
@@ -192,8 +190,26 @@ static String getPathInContext(Request request)
+     * which is appropriate for wall clock time.</p>
+     * @return The timestamp that the request was received/created in nanoseconds, relative to {@link System#nanoTime()}.
+     */
+    default long getNanoTimeStamp()
```
Rename to `getNanoTime()`.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Request.java:
```diff
@@ -192,8 +190,26 @@ static String getPathInContext(Request request)
+     */
+    default long getNanoTimeStamp()
+    {
+        return NanoTime.now() - TimeUnit.MILLISECONDS.toNanos(System.currentTimeMillis() - getTimeStamp());
```
I don't think this implementation is correct, since `currentTimeMillis()` can drift.
Must use another field to store the creation/received nanoTime.

gregw replied:
@sbordet we do. It is in the HttpStream.  This is just used for test code.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:
```diff
@@ -921,6 +870,16 @@ public long getTimeStamp()
+            if (stream != null)
+                return stream.getNanoTimeStamp();
+
+            return NanoTime.now() - TimeUnit.MILLISECONDS.toNanos(System.currentTimeMillis() - getTimeStamp());
```
Should call `super` instead?
As we don't want to duplicate the (wrong) logic here.

gregw replied:
Does super call a default implementation? I don't think so.

sbordet reviewed (approved):
LGTM. I have a few nits, but I'll fix them myself.

sbordet committed:
Cleaned up Request.getNanoTime().
Renamed HttpStream.getNanoTimeStamp() to getNanoTime().

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Merged branch 'jetty-12.0.x' into 'jetty-12-handler-as-boolean-processor'.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

lorban reviewed (approved):
LGTM

sbordet committed:
Avoid using an extra field to store nanoTime.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Now throw from getNanoTime() if there is no stream.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>
